packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,204 @@
+      score = (0.75 * x) / (1 + x);
+      if (isSelfPayment(transaction,client)) {
+        return score * DENIABILITY_FACTOR;
+      }
+    }
+  }
+  return score;
+}
+
+/*
+Determine whether a Bitcoin transaction is a self-payment, meaning the sender and recipient
+are the same entity. To check this, we make an RPC call to the watcher wallet asking for the
+amount that a given address holds. If the call returns a number then it is part of wallet
+otherwise it is not a self payment.
+*/
+function isSelfPayment(transaction: any, client: BlockchainClient): boolean {
+  transaction.vout.forEach(async op => {
+     if(await client.getAddressStatus(op.scriptPubKey.address) === undefined){
+       return false;
+     }
+  })
+  return true;
+}
+
+/* TODO : replace any type to custom types for Transactions and UTXOs*/
+/* 
+In order to score for address reuse we can check the amount being hold by reused UTXOs 
+with respect to the total amount
+*/
+export function addressReuseFactor(utxos: Array<any>): number {
+  let reused_amount : number = 0;
+  let total_amount : number = 0;
+  utxos.forEach((utxo) => {
+    if (utxo.reused) {
+      reused_amount += utxo.amount;
+    }
+    total_amount += utxo.amount;
+  });
+  return reused_amount / total_amount;
+}
+
+/*
+If we are making payment to other wallet types then the privacy score should decrease because 
+the change received will be at address of our wallet type and it will lead to derivation that 
+we still own that amount.
+*/
+export function addressTypeFactor(transactions : Array<any>, walletAddressType : string): number {
+    let P2WSH : number = 0;
+    let P2PKH : number = 0;
+    let P2SH : number = 0;
+    let atf : number = 1;
+    transactions.forEach(tx => {
+        tx.vout.forEach(output => {
+            let address = output.scriptPubKey.address;
+            if (address.startsWith("bc")) {
+                //Bech 32 Native Segwit (P2WSH) or Taproot
+                P2WSH += 1;
+            } else if (address.startsWith("1")) {
+                // Legacy (P2PKH)
+                P2PKH += 1;
+            } else {
+                // Segwith (P2SH)
+                P2SH += 1;
+            }
+        })
+    });
+
+    if (walletAddressType == "P2WSH" && (P2WSH != 0 && (P2SH != 0 || P2PKH != 0))) {
+        atf = 1 / (P2WSH + 1);
+    } else if (walletAddressType == "P2PKH" && (P2PKH != 0 && (P2SH != 0 || P2WSH != 0))) {
+        atf = 1 / (P2PKH + 1);
+    } else if (walletAddressType == "P2SH" && (P2SH != 0 && (P2WSH != 0 || P2PKH != 0))) {
+        atf = 1 / (P2SH + 1);
+    } else {
+        atf = 1;
+    }
+    return atf;
+}
+
+/* 
+The spread factor using standard deviation helps in assessing the dispersion of UTXO values.
+In Bitcoin privacy, spreading UTXOs reduces traceability by making it harder for adversaries
+to link transactions and deduce the ownership and spending patterns of users.
+*/
+export function utxoSpreadFactor(utxos : Array<any>) : number {
```

Comment:
so this would be all outputs for all transactions or just the current unspent outputs of the transaction?

Reply:
If the former maybe looping through a list of all historical txs is desirable?

Reply:
Nope. It is latter, "just the current unspent outputs of the transaction".

Reply:
oh wait, utxos of a tx or wallet? wouldn't the spread factor of past utxos also impact privacy? Or is that captured elsewhere

Reply:
"UTXOs sitting in the wallet at the moment". Hmm, if the goal is to account for UTXOs sitting in past then I think we are not capturing it yet. 

Privacy is being analysed on number of inputs and outputs which already accounts for past spend history and takes into account of how UTXOs were spent. This specific part of code accounts for how well your wallet is balanced with UTXOs to be spent in the future. 

Past UTXO Set [0.5, 0.5 BTC]
If both were consolidated, then 1 BTC UTXO still remains. based on the logic of input and output we get the idea that there was a consolidation happened in past and will degrade the privacy in math. But what the UTXO Spread factor talks about is how well your current folio is [1 BTC] to spend good transactions and improvises the privacy.

Reply:
Ok yeah. I think we've talked about this before and I just lost track. Might be worth a mention in the actual blog post in this section that this ignores historical factors as they are captured in specific other metrics, which i believe is what you're saying right?

Reply:
Yes